java/src/jmri/Block.java

```diff
@@ -318,7 +318,32 @@ public void setState(int v) {
         int old = _current;
         _current = v;
         // notify
-        firePropertyChange("state", old, _current);
+        // It is rather unpleasant that the following needs to be done in a try-catch, but exceptions have been observed: 
+        // java.lang.IllegalArgumentException: illegal component position
+	// at java.awt.Container.addImpl(Container.java:1087)
+	// at javax.swing.JLayeredPane.addImpl(JLayeredPane.java:231)
+	// at java.awt.Container.add(Container.java:973)
+	// at jmri.jmrit.display.Editor$TargetPane.add(Editor.java:608)
+	// at jmri.jmrit.display.Editor.addToTarget(Editor.java:1716)
+	// at jmri.jmrit.display.Editor.displayLevelChange(Editor.java:1673)
+	// at jmri.jmrit.display.PositionableJComponent.setDisplayLevel(PositionableJComponent.java:143)
+	// at jmri.jmrit.display.controlPanelEditor.shape.PositionableShape.<init>(PositionableShape.java:70)
+	// at jmri.jmrit.display.controlPanelEditor.shape.PositionableRectangle.<init>(PositionableRectangle.java:16)
+	// at jmri.jmrit.display.controlPanelEditor.shape.PositionableRoundRect.<init>(PositionableRoundRect.java:20)
+	// at jmri.jmrit.display.controlPanelEditor.shape.LocoLabel.<init>(LocoLabel.java:13)
+	// at jmri.jmrit.display.IndicatorTrackPaths.setLocoIcon(IndicatorTrackPaths.java:132)
+	// at jmri.jmrit.display.IndicatorTrackIcon.setStatus(IndicatorTrackIcon.java:282)
+	// at jmri.jmrit.display.IndicatorTrackIcon.propertyChange(IndicatorTrackIcon.java:263)
+	// at java.beans.PropertyChangeSupport.fire(PropertyChangeSupport.java:335)
+	// at java.beans.PropertyChangeSupport.firePropertyChange(PropertyChangeSupport.java:327)
+	// at java.beans.PropertyChangeSupport.firePropertyChange(PropertyChangeSupport.java:263)
+	// at jmri.implementation.AbstractNamedBean.firePropertyChange(AbstractNamedBean.java:243)
+	// at jmri.Block.setState(Block.java:321)
```

Comment:
When you see things like this, it's good to capture a bit more of the trace (all the way back to the origination) and the thread name to help track down the cause.  Generally, they're caused by something happening _off_ the GUI+Layout thread, although sometimes that doesn't bit right away.

To help with the debugging, it would be good to have the log.debug(..) below be a log.warn(..) so we can't just ignore these. If there are too many, you can use a warnOnce.

Reply:
I see that I have used log.debug once and log.error the other two places. Would it be better if I replace all three with log.warn?  Or is log.error better?

And should I add a logging of the full stack trace for the exception like this:
private void logStackTrace(Exception e) {
         StringWriter outError = new StringWriter();
         e.printStackTrace(new PrintWriter(outError));
         log.error(outError.toString());
 }

How can I add the thread name?

I don’t think there will be problems with too many such incidents. They seem to be quite sporadic.

Reply:
Error carries the idea that something is broken internally, so perhaps error is better for all three.

Adding the exception as an additional parameter at the end of the log.debug(..) logs the _entire_ exception, so when it's printed the stack trace and thread name is added.

```
log.debug(" {} got exception during fireProperTyChange({},{})", getDisplayName(), old, _current, e);
```